modules/effects/arrow-effects/src/main/kotlin/arrow/effects/IORace.kt

```diff
@@ -0,0 +1,114 @@
+package arrow.effects
+
+import arrow.core.*
+import arrow.effects.internal.IOConnection
+import arrow.effects.internal.asyncContinuation
+import arrow.effects.internal.unsafe
+import arrow.effects.typeclasses.Fiber
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.startCoroutine
+import kotlin.coroutines.suspendCoroutine
+
+typealias Pair<A, B> = Either<Tuple2<A, Fiber<ForIO, B>>, Tuple2<Fiber<ForIO, A>, B>>
+
+fun <A, B> IO.Companion.racePair(ctx: CoroutineContext, ioA: IOOf<A>, ioB: IOOf<B>): IO<Pair<A, B>> =
+  IO.async { conn, cb ->
+    val active = AtomicBoolean(true)
+
+    // Cancelable connection for the left value
+    val connA = IOConnection()
+    val promiseA = Promise.unsafe<Either<Throwable, A>>()
+
+    // Cancelable connection for the right value
+    val connB = IOConnection()
+    val promiseB = Promise.unsafe<Either<Throwable, B>>()
+
+    conn.pushPair(connA, connB)
+
+    val a: suspend () -> A = {
+      suspendCoroutine { ca: Continuation<A> ->
+        IORunLoop.startCancelable(ioA.fix(), connA) { either: Either<Throwable, A> ->
+          either.fold({ error ->
+            ca.resumeWith(Result.failure(error))
+          }, { a ->
+            ca.resumeWith(Result.success(a))
+          })
+        }
+      }
+    }
+
+    val b: suspend () -> B = {
+      suspendCoroutine { ca: Continuation<B> ->
+        IORunLoop.startCancelable(ioB.fix(), connB) { either: Either<Throwable, B> ->
+          either.fold({ error ->
+            ca.resumeWith(Result.failure(error))
+          }, { b ->
+            ca.resumeWith(Result.success(b))
+          })
+        }
+      }
+    }
+
+    a.startCoroutine(asyncContinuation(ctx) { either ->
+      either.fold({ error ->
+        if (active.getAndSet(false)) { //if an error finishes first, stop the race.
+          connB.cancel().fix().unsafeRunAsync { r2 ->
```

Comment:
Nit: in the past I've been bitten by `cancel()` not immediately triggering a cancellacion. Is it worth renaming it to something that suggest it's returning an object you have to act one?

Reply:
I'm unsure, to be honest. That's why I liked the property based API because it was giving me a false feeling that things still needed to be run 😅

We do this all over tho. Same for `Fiber` with suggested review changes. To be fair I don't I don't have a strong opinion.